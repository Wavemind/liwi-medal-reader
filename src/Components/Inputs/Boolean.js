/**
 * The external imports
 */
import React, { useEffect, useState } from 'react'
import { View, TouchableOpacity, Text } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'

/**
 * The internal imports
 */
import { useTheme } from '@/Theme'
import { translate } from '@/Translations/algorithm'
import { getYesAnswer, getNoAnswer } from '@/Utils/Answers'
import DefineType from '@/Store/Modal/DefineType'
import ToggleVisbility from '@/Store/Modal/ToggleVisibility'
import SetAnswer from '@/Store/MedicalCase/SetAnswer'

const Boolean = ({ questionId, emergency, disabled = false }) => {
  // Theme and style elements deconstruction
  const {
    Components: { booleanButton },
    Layout,
  } = useTheme()
  const dispatch = useDispatch()

  // Get node from algorithm
  const question = useSelector(
    state => state.medicalCase.item.nodes[questionId],
  )
  const currentNode = useSelector(
    state => state.algorithm.item.nodes[question.id],
  )

  // Local state definition
  const [value, setValue] = useState(question.answer)

  const yesAnswer = getYesAnswer(currentNode)
  const noAnswer = getNoAnswer(currentNode)

  /**
   * Set node answer and handle emergency action
   * @param {integer} answerId
   */
  const setAnswer = async answerId => {
    if (emergency && yesAnswer.id === answerId) {
      await dispatch(DefineType.action({ type: 'emergency' }))
      await dispatch(ToggleVisbility.action({}))
    }
    setValue(answerId)
  }

  /**
   * Update value in store when value changes
   */
  useEffect(() => {
    if (question.value !== value) {
      dispatch(SetAnswer.action({ nodeId: question.id, value }))
    }
  }, [value])

  return (
    <View style={Layout.row}>
      <View
        key="booleanButton-left"
        style={booleanButton.buttonWrapper(
          'left',
          value === yesAnswer.id,
          disabled,
        )}
      >
        <TouchableOpacity
          style={Layout.center}
          onPress={() => setAnswer(yesAnswer.id)}
          disabled={disabled}
        >
          <Text style={booleanButton.buttonText(value === yesAnswer.id)}>
            {translate(yesAnswer.label)}
          </Text>
        </TouchableOpacity>
      </View>
      <View
        key="booleanButton-right"
        style={booleanButton.buttonWrapper(
          'right',
          value === noAnswer.id,
          disabled,
        )}
      >
        <TouchableOpacity
          style={Layout.center}
          onPress={() => setAnswer(noAnswer.id)}
          disabled={disabled}
        >
          <Text style={booleanButton.buttonText(value === noAnswer.id)}>
            {translate(noAnswer.label)}
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  )
}

export default Boolean
