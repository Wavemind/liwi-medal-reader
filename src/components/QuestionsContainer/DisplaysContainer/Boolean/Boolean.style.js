import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  bottomInput: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    margin: 0,
    padding: 0,
  },
});
