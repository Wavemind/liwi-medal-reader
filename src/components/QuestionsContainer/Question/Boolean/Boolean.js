// @flow

import * as React from 'react';
import { Button, H2, Icon, Text, View, Grid, Col } from 'native-base';
import type { NavigationScreenProps } from 'react-navigation';
import type { SessionsProviderState } from 'engine/contexts/Sessions.context';
import { getSession } from 'engine/api/LocalStorage';
import { QuestionView, RightView } from '../../../../template/layout';
import { liwiColors } from '../../../../utils/constants';

type Props = NavigationScreenProps & {};

type State = {};

export default class Boolean extends React.PureComponent<Props, State> {
  state = {
    answer: null,
  };

  _handleNo = () => {
    let newAnswer;
    switch (this.state.answer) {
      case true:
        newAnswer = false;
        break;
      case false:
        newAnswer = null;
        break;
      case null:
        newAnswer = false;
        break;
    }

    this.setState({ answer: newAnswer });
  };

  _handleYes = () => {
    let newAnswer;
    switch (this.state.answer) {
      case true:
        newAnswer = null;
        break;
      case false:
        newAnswer = true;
        break;
      case null:
        newAnswer = true;
        break;
    }
    this.setState({ answer: newAnswer });
  };

  render() {
    const { question, styles } = this.props;
    const { answer } = this.state;

    return (
      <Grid>
        <Col>
          <Button
            square
            onPress={this._handleYes}
            style={
              answer === true
                ? { backgroundColor: '#c0c0c0' }
                : { backgroundColor: liwiColors.whiteColor }
            }
          >
            <Icon
              style={
                answer
                  ? { color: liwiColors.greenColor }
                  : { color: liwiColors.blackColor }
              }
              name={'check'}
              type={'MaterialCommunityIcons'}
            />
          </Button>
        </Col>
        <Col>
          <Button
            square
            onPress={this._handleNo}
            style={
              answer === false
                ? { backgroundColor: '#c0c0c0' }
                : { backgroundColor: liwiColors.whiteColor }
            }
          >
            <Icon
              style={
                answer === false
                  ? { color: liwiColors.redColor }
                  : { color: liwiColors.blackColor }
              }
              name={'cross'}
              type={'Entypo'}
            />
          </Button>
        </Col>
      </Grid>
    );
  }
}
