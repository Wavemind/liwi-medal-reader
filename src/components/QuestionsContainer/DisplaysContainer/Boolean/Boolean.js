// @flow

import * as React from 'react';
import { Button, Text, View } from 'native-base';
import { Dimensions } from 'react-native';

import { LeftButton, RightButton } from '../../../../template/layout';
import { categories, modalType } from '../../../../../frontend_service/constants';
import { liwiColors } from '../../../../utils/constants';
import { styles } from './Boolean.style';
import { translateText } from '../../../../utils/i18n';

export default class Boolean extends React.Component {
  static defaultProps = {
    question: {},
    category: null,
  };

  state = {
    value: null,
  };

  shouldComponentUpdate(nextProps, nextState) {
    const { value } = this.state;
    const { question } = this.props;
    return nextState.value !== value || question.answer !== nextProps.question.answer;
  }

  componentDidMount() {
    const {
      app: { algorithm },
      question,
      question: { answer },
    } = this.props;

    const currentNode = algorithm.nodes[question.id];

    const idYes = Number(Object.keys(currentNode.answers)[0]);
    const idNo = Number(Object.keys(currentNode.answers)[1]);

    if (answer === idYes) {
      this.setState({ value: true });
    } else if (answer === idNo) {
      this.setState({ value: false });
    }
  }

  /**
   * Set answer in store
   * @param {boolean} answer - Value to save in store
   * @returns {null}
   * @private
   */
  _handleClick = async (answer) => {
    const {
      app: { algorithm, set },
      question,
      setAnswer,
      setPatientValue,
      patientValueEdit,
      updateModalFromRedux,
    } = this.props;
    const currentNode = algorithm.nodes[question.id];

    const idYes = Number(Object.keys(currentNode.answers)[0]);
    const idNo = Number(Object.keys(currentNode.answers)[1]);

    let newAnswer = Number(answer);

    if (newAnswer === idYes) {
      this.setState({ value: true });
    } else if (newAnswer === idNo) {
      this.setState({ value: false });
    }
    // Give the time to the component to render the view before updating the state
    await new Promise((resolve) => setTimeout(resolve, 25));

    // Break if complaintCategory
    if (currentNode.category === categories.complaintCategory && answer === question.answer) {
      return null;
    }

    if (patientValueEdit) {
      setPatientValue(question.id, newAnswer);
    } else {
      setAnswer(algorithm, question.id, newAnswer);
    }

    set('answeredQuestionId', question.id);

    // Open emergency modal
    if (currentNode?.emergency_status === 'emergency' && newAnswer === idYes) {
      updateModalFromRedux({}, modalType.emergencyWarning);
    }
  };

  // TODO: Sorry for this. But I don't have any strength to refactor this shit
  render = () => {
    const {
      app: { t, algorithm, algorithmLanguage },
      question,
      index,
      isReadOnly,
    } = this.props;
    const { value } = this.state;
    const { answers, label, category } = algorithm.nodes[question.id];

    // Define the id for the answer
    const idYes = Number(Object.keys(answers)[0]);
    const idNo = Number(Object.keys(answers)[1]);

    const margin = 15;
    const sizeButton = Math.floor(Dimensions.get('window').width / 3 - margin * 2.7);
    const mod = index % 3;

    let styleMargin = {};
    let activeStyle;
    let idOnPress;
    let concatStyle;

    let RenderJsx;

    // By the type of boolean node
    if (category === categories.complaintCategory && !isReadOnly) {
      // If this is a chief complain
      if (mod === 0) {
        styleMargin = {
          marginTop: margin,
          marginBottom: 0,
          marginLeft: margin,
          marginRight: 0,
        };
      } else if (mod === 1) {
        styleMargin = {
          marginTop: margin,
          marginBottom: 0,
          marginLeft: margin,
          marginRight: margin,
        };
      } else if (mod === 2) {
        styleMargin = {
          marginTop: margin,
          marginBottom: 0,
          marginLeft: 0,
          marginRight: margin,
        };
      }

      if (value === null) {
        activeStyle = { backgroundColor: 'transparent' };
        idOnPress = idYes;
      } else {
        idOnPress = value ? idYes : idNo;
        activeStyle = {
          borderColor: liwiColors.darkerGreyColor,
        };
      }

      concatStyle = {
        width: sizeButton,
        flexDirection: 'column',
        height: sizeButton,
        elevation: 5,
        borderRadius: 2,
        borderColor: liwiColors.darkGreyColor,
        ...styleMargin,
        ...activeStyle,
        justifyContent: 'space-between',
        backgroundColor: liwiColors.whiteColor,
        borderWidth: 2,
        paddingTop: 0,
        paddingBottom: 0,
      };

      // Only one button for this type of node
      RenderJsx = () => (
        <Button key={index + category} onPress={() => this._handleClick(idOnPress)} style={concatStyle} light>
          <View complaintCategories>
            <Text center size-auto>
              {translateText(label, algorithmLanguage)}
            </Text>
          </View>
          <View style={styles.bottomInput}>
            <LeftButton active={value} onPress={() => this._handleClick(idYes)} disabled={isReadOnly}>
              <Text white={value} center>
                {t('question:yes')}
              </Text>
            </LeftButton>

            <RightButton active={value === false} onPress={() => this._handleClick(idNo)} disabled={isReadOnly}>
              <Text center white={value === false}>
                {t('question:no')}
              </Text>
            </RightButton>
          </View>
        </Button>
      );
    } else {
      const leftLabel = translateText(Object.values(answers)[0].label, algorithmLanguage);
      const rightLabel = translateText(Object.values(answers)[1].label, algorithmLanguage);

      RenderJsx = () => (
        <View answer>
          <LeftButton active={value} onPress={() => this._handleClick(idYes)} disabled={isReadOnly}>
            <Text white={value} center style={{ fontSize: leftLabel.length > 3 ? 10 : 16 }}>
              {leftLabel}
            </Text>
          </LeftButton>
          <RightButton active={value === false} onPress={() => this._handleClick(idNo)} disabled={isReadOnly}>
            <Text center white={value === false} style={{ fontSize: rightLabel.length > 3 ? 10 : 16 }}>
              {rightLabel}
            </Text>
          </RightButton>
        </View>
      );
    }

    return <RenderJsx />;
  };
}
