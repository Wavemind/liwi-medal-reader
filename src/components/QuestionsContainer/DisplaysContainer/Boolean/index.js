import { connect } from 'react-redux';
import Boolean from './Boolean';
import { withApplication } from '../../../../engine/contexts/Application.context';
import { setPatientValue, setAnswer } from '../../../../../frontend_service/actions/creators.actions';

const mapStateToProps = (medicalCase) => {
  return { medicalCase };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    setAnswer: (index, value) => dispatch(setAnswer(index, value)),
    setPatientValue: (index, value) => dispatch(setPatientValue(index, value)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(withApplication(Boolean));
