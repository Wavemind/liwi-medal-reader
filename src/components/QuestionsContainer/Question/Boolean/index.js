import Boolean from './Boolean';

export default Boolean;
