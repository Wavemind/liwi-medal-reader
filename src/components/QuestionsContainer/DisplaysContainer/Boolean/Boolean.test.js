import * as React from 'react';
import Boolean from './Boolean';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
// setup file
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
/*
 * https://medium.com/@AidThompsin/heres-how-you-unit-test-textinput-with-react-native-63e1f7692b17
 * https://www.freecodecamp.org/news/components-testing-in-react-what-and-how-to-test-with-jest-and-enzyme-7c1cace99de5/
 * */

test('renders patient list correctly', () => {
  const tree = renderer
    .create(
      <Boolean
        question={{
          id: 47,
          type: 'Question',
          reference: 'V1',
          description: null,
          label: 'MMR: 2 doses completed',
          answer: null,
          answers: {
            '91': {
              id: 91,
              reference: 'V1_1',
              label: 'Yes',
              value: null,
              operator: null,
            },
            '92': {
              id: 92,
              reference: 'V1_2',
              label: 'No',
              value: null,
              operator: null,
            },
          },
          category: 'vaccine',
          counter: 0,
          dd: [{ id: 16, conditionValue: false }],
          display_format: 'RadioButton',
          priority: 'basic',
          qs: [],
          value: 0,
          value_format: 'Boolean',
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
