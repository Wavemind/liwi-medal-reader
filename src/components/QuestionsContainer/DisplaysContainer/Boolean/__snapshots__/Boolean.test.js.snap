// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders patient list correctly 1`] = `
<View
  answer={true}
  style={Object {}}
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      center={true}
      style={
        Object {
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      uppercase={false}
      white={false}
    >
      Oui
    </Text>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      center={true}
      style={
        Object {
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      uppercase={false}
      white={false}
    >
      Non
    </Text>
  </View>
</View>
`;
